src/mesh/Mesh.ts

```diff
@@ -0,0 +1,291 @@
+    this.onConnectionElements = [];
+    this.self = new Peer(options.host, options.port, uuid());
+    this.knownPeers = new KnownPeers(this.self);
+
+    this.logger.info('MY ID', this.self.id);
+
+    if (options.certificate && options.privateKey) {
+      this.server = https.createServer({
+        key: options.privateKey,
+        cert: options.certificate
+      });
+    } else {
+      this.server = http.createServer();
+    }
+
+    this.serverSocket = new Server<ServerToClientEvents>(this.server);
+
+    const otherPeers = options.knownPeers || [] as Peer[];
+
+    this.server.listen(this.self.port, () => {
+      this.logger.info('Server started on port', this.self.port);
+    });
+
+    this.clients = {};
+    if (otherPeers.length) {
+      this.connectNewPeers(otherPeers);
+    }
+
+    this.setupServerListeners();
+  }
+
+  private setupServerListeners() {
+    this.serverSocket.on('connection', (socket: ServerSocket<ClientToServerEvents, ServerToClientEvents>) => {
+
+      this.logger.info('on server connection');
+
+      // setup default events
+      socket.emit('getInfo', (info: Info) => {
+        this.logger.info('getInfo cb');
+
+        if (this.password) {
+          if (!info.password || info.password !== this.password) {
+            return socket.disconnect(true);
+          }
+        }
+
+        if (this.state !== 'connection') {
+          this.newState('connection');
+        }
+
+        this.logger.info('server connect new clients', info.knownPeers);
+
+        let newPeer = info.self;
+
+        if (!newPeer.host || newPeer.host === "localhost") {
+          let remoteAddress = socket.client.conn.remoteAddress.split(':');
+          newPeer.host = remoteAddress[remoteAddress.length - 1]; // TODO: check is it needed
+        }
+
+        this.knownPeers.add(newPeer);
+        this.connectNewPeers(info.knownPeers.peers);
+
+        this.clients[info.self.id] = socket;
+        this.logger.info('server connect new clients', Object.keys(this.clients));
+
+        socket.emit('sendPeers', {
+          peers: this.knownPeers.peers,
+          self: this.self
+        });
+      });
+
+      socket.on('joined', () => {
+        this.onConnectionElements.pop();
+
+        this.logger.info('on joined', this.onConnectionElements.length);
+
+        if (this.onConnectionElements.length && this.state !== 'joined') {
+          this.newState('joined');
+        }
+      });
+
+      socket.on('disconnect', () => {
+        const key = this.getKeyByValue(this.clients, socket);
+        delete this.clients[key];
+
+        this.logger.info('client', key, 'was disconnected, clients left', Object.keys(this.clients));
+      });
+
+      //  custom events
+      for (let on of this.customEvents) {
+        this.logger.info('add listener', on.name);
+        // cast to never because should subscribe on all custom events but cannot get all custom event data
+        // TODO: or can? Add generic data to the class like socket.io?
```

Comment:
@xziy What do you think about such feature? Is it good to do it? Maybe with do it as optional opportunity?